File: cognipy/cognipylib/cognipy/cnl/dl/CalcDynamicDepths.java
package cognipy.cnl.dl;

import cognipy.*;
import cognipy.cnl.*;
import java.util.*;

public class CalcDynamicDepths extends GenericVisitor
{
	public HashSet<Integer> IntersectionDepth = new HashSet<Integer>();
	public HashSet<Integer> UnionDepth = new HashSet<Integer>();
	public HashSet<Integer> HasKeyDepth = new HashSet<Integer>();

	@Override
	public Object Visit(ConceptAnd e)
	{
		IntersectionDepth.add(e.Exprs.size());
		return super.Visit(e);
	}

	@Override
	public Object Visit(ConceptOr e)
	{
		UnionDepth.add(e.Exprs.size());
		return super.Visit(e);
	}
	@Override
	public Object Visit(HasKey e)
	{
		UnionDepth.add(e.DataRoles.size() + e.Roles.size());
		return super.Visit(e);
	}
}

File: docsrc/jupyter_notebooks/BugReports/cognipy/CogniPyLib/JenaRules/CalcDynamicDepths.cs
using System.Collections.Generic;

namespace CogniPy.CNL.DL
{
    internal class CalcDynamicDepths : GenericVisitor
    {
        public HashSet<int> IntersectionDepth = new HashSet<int>();
        public HashSet<int> UnionDepth = new HashSet<int>();
        public HashSet<int> HasKeyDepth = new HashSet<int>();

        public override object Visit(ConceptAnd e)
        {
            IntersectionDepth.Add(e.Exprs.Count);
            return base.Visit(e);
        }

        public override object Visit(ConceptOr e)
        {
            UnionDepth.Add(e.Exprs.Count);
            return base.Visit(e);
        }
        public override object Visit(HasKey e)
        {
            UnionDepth.Add(e.DataRoles.Count + e.Roles.Count);
            return base.Visit(e);
        }
    }
}